### ISSUE TITLE
AIRR format output

### ISSUE THREAD

**Issue by leeanapeters** [liulab-dfci/TRUST4#113]
Hi, thank you for the fantastic package!

I initially ran TRUST4 with an earlier version that I guess did not output in an airr format. I know you can convert the existing barcode report to this format but I am having issues and wanted to confirm I am doing the correct thing. 

I ran this 
perl /orange/brusko/single_cell/NS1998/TRUST4/TRUST4/trust-airr.pl /orange/brusko/single_cell/NS1998/src/TRUST_6482_GE_S2_L001_R1_001_annot.fa AIRR_report_6482

but this failed with no error.

I had to re-run the whole sample with the newer version to obtain the airr report. 

Any guidance on how I can convert the existing format would be much appreciated!

Thanks

Leeana

**Comment by leeanapeters** [comment]
When I run that I get empty output. This is single cell data so I used the barcode_report.tsv file. Not sure if that makes a difference

**Comment by mourisl** [comment]
For single-cell data, you can run the command like 
"perl trust-airrp.pl trust_barcode_report.tsv trust_annot.fa --format barcoderep > AIRR_report.tsv".

**Comment by leeanapeters** [comment]
Thank you! I am still having some issues but it seems to be coming from the trust-airrp.pl script, I get this error: 

readline() on closed filehandle FP

**Comment by mourisl** [comment]
What is the earlier version of TRUST4 you used? Does the perl warning give you the line number of which it occurs? Thanks.

**Comment by leeanapeters** [comment]
I believe 1.0.6? I installed the version from github just a few days ago for this run as the earlier version that I had did not have a airr.pl script.

The full error is 
readline() on closed filehandle FP at /orange/brusko/single_cell/NS1998/TRUST4/TRUST4/trust-airr.pl line 166.


Thank you so much for the help!

**Comment by sciencepeak** [comment]
I saw the usage statement like this in the perl script.
usage: trust-airr.pl trust_report.tsv trust_annot.fa [--format simplerep|cdr3|barcoderep]> trust_airr.tsv

Could you please update the readme file or the trust-airr.pl file so that the --format three options are explained. I am confused by the meaning of the three --format options. From the previous conversion in this post, I can only infer that the "barcoderep" is for the single cell data.

Thanks a lot.

**Comment by mourisl** [comment]
I will add more documentation to this script.

Essentially "simplerep" is the default,original simple report output file from TRUST4, cdr3 is the "trust_cdr3.out" file and barcoderep is the barcode report file.

**Comment by sciencepeak** [comment]
Cool. I am curious about that I can not see a cdr3 column in the airr-format output. I only see cdr1 and cdr2 column.
I am not very familar with the airr format, but I know the importance of the cdr3 region.

**Comment by mourisl** [comment]
AIRR format uses "junction" and "junction_aa" field to represent CDR3. In AIRR, junction means CDR3 plus the two flanking conserved motif amino acid. "Junction" field is mandatory in AIRR, while "CDR3" is optional.

**Comment by sciencepeak** [comment]
Thanks for your great explanation.

**Comment by sciencepeak** [comment]
Hi, I manully run the trust-airr.pl with --format cdr3. I don't know why the result table has much less rows than the table derived from the default --format simplerep that are automatically generated by the run-trust4.

As you suggested, the --format cdr3 probably reports from xxx_cdr3.out, the --format remplerep probably reports from xxx_final.out. I still don't know the difference of these two files and how to choose between them for the downstream analysis such as SCOPer.

Thanks a lot for your work.

**Comment by mourisl** [comment]
The cdr3 file contains many partial CDR3s, which are filtered in the final trust_report.tsv file. 

I would recommend using the trust-airr.pl script on the trust_report.tsv file (--format simplerep). The simple report file contains some other postprocessings.

**Comment by sciencepeak** [comment]
Cool. Now I understand how to run the trust-airr.pl script manually. And I understand your recommendation on --format simplerep.

```shell
perl ./TRUST4/trust-airr.pl ./trust_result/TRUST_xxx_report.tsv ./trust_result/TRUST_xxx_annot.fa --format simplerep > ./trust_result/trust_airr.converted.report.tsv

perl ./TRUST4/trust-airr.pl ./trust_result/TRUST_xxx_cdr3.out ./trust_result/TRUST_xxx_annot.fa --format cdr3 > ./trust_result/trust_airr.converted.cdr3.tsv
```

Since the --format simplerep is the default, I can either run the whole run-trust4 command line or manually run the trust-airr.pl to get the airr format output.

As for the the fields "sequence_alignment" and "germline_alignment" in SCOPer, I will explore more to see whether the two fields are essential for the function of SCOPer.

Thanks a lot.

**Comment by sciencepeak** [comment]
As I read the [SCOPer paper](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007977#sec002) and [online documentation](https://scoper.readthedocs.io/en/stable/vignettes/Scoper-Vignette/), I found that the fields "sequence_alignment" and "germline_alignment" are used, at least in [one](https://scoper.readthedocs.io/en/stable/topics/spectralClones/) of the [three methods](https://scoper.readthedocs.io/en/stable/vignettes/Scoper-Vignette/) in SCOPer. Then, I tried to use the airr-format output from TRUST4 as input of SCOPer. The SCOPer can run and give normal results, using any of the three methods. This is a good news. It seems that the SCOPer claimed to require the two fields "sequence_alignment" and "germline_alignment", but actually the two fields are unnecessary.

As I read the document of [airr-format standard](https://docs.airr-community.org/en/stable/datarep/rearrangements.html), I found that the "sequence alignment" is relatively easy to derive from the perspective of TRUST4, whereas "germline_alignment" needs more complicated calculation and inferrence that might be beyond the current purpose of TRUST4. Am I right?

Thanks a lot for your inspiring discussion.

**Comment by mourisl** [comment]
Yes, the sequence alignment can be easily inferred from the v_cigar and j_cigar. I don't think the germline_alignment can be directly inferred from the current TRUST4 AIRR output. 

To infer the germline sequence, we can run the annotator with option --geneAlignment to output the alignment detail between the sequence and IMGT reference. The output is clumsy, and I'm thinking of a better way to present these results.

**Comment by snowGoose-Chen** [comment]
Could you please explain how to infer sequence alignment from the v_cigar and j_cigar?

**Comment by mourisl** [comment]
If V cigar starts like xN, then you can skip the first x bases from the contig. If J cigar ends like xN, you can remove the last x bases from the contig. The remaining sequence will be sequence alignment.

**Comment by snowGoose-Chen** [comment]
Thanks a lot for your work and your timely reply!

**Comment by insidemymind** [comment]
Hi. Thanks for this wonderful tool! 

Recent Trust4 airr format output has sequence_alignment & germline_alignment, but the perl script "trust-airr.pl" is not capable of those two output fields from previous trust4 version (1.0.4). To get two fields I migth have to re-run all the samples by recent version of Trust4. So is there a way to bypass this problem?

**Comment by mourisl** [comment]
If you have the intermediate files, like _final.out, _assembled_reads.fa, you can rerun TRUST4 with the option "--stage 2", which should be much faster to rerun.

**Comment by insidemymind** [comment]
Thanks, I will try that.



---- Replied Message ----
| From | Li ***@***.***> |
| Date | 01/26/2024 22:40 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [liulab-dfci/TRUST4] AIRR format output (Issue #113) |

If you have the intermediate files, like _final.out, _assembled_reads.fa, you can rerun TRUST4 with the option "--stage 2", which should be much faster to rerun.
